[GTK][WPE][OpenXR] Expose a way to query the immersive session status

spenap opened the pull request:
#### d4e9102d6bb9aeac9df1844b949306945884d9ec
<pre>
[GTK][WPE][OpenXR] Expose a way to query the immersive session status
<a href="https://bugs.webkit.org/show_bug.cgi?id=297048">https://bugs.webkit.org/show_bug.cgi?id=297048</a>

Reviewed by Carlos Garcia Campos and Adrian Perez de Castro.

A new property &quot;is-immersive-mode-enabled&quot; is added to WebKitWebView,
allowing clients to query its value and to subscribe to its notify
signal.

The property is synchronised with the OpenXR coordinator&apos;s entering and
exiting the session, where the WebPageProxy&apos;s UI client is used to
report back to the WebView.

This renames webkit_web_view_end_immersive_session into
webkit_web_view_leave_immersive_mode for better consistency.

* Source/WebKit/UIProcess/API/APIUIClient.h: two new methods to report
  entering and exiting the XR session.
(API::UIClient::didStartXRSession):
(API::UIClient::didEndXRSession):
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp: implementing the
  UI client methods in terms of updating the immersive mode enabled
property.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: add the new
  property, rename method leaving the mode.
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in: rename method
  leaving the mode.
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h: adding method
  setting the immersive mode enabled property.
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp: report when we
  enter and leave the XR session.
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
* Tools/MiniBrowser/wpe/main.cpp: query whether we&apos;re in immersive mode
  before attempting to leave.

Canonical link: <a href="https://commits.webkit.org/298568@main">https://commits.webkit.org/298568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22b31dc6a3341898a24079a52134c7ea89877e5b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35581 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26117 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121976 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66458 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117816 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36270 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44168 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88062 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118875 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28939 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104030 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68475 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28061 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22138 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65646 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98327 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22272 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125128 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42814 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32132 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96816 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43180 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100218 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96603 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24572 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41864 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19736 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38738 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42703 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48292 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42170 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45504 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43877 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -998,6 +1000,7 @@ static void webkitWebViewConstructed(GObject* object)
     priv->backForwardList = adoptGRef(webkitBackForwardListCreate(&getPage(webView).backForwardList()));
     priv->windowProperties = adoptGRef(webkitWindowPropertiesCreate());
     priv->isWebProcessResponsive = true;
+    priv->hasImmersiveSession = false;
```
This should not be needed, new instances are zero-initialized so this will be `false` already because of that.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1748,6 +1754,20 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
         WEBKIT_PARAM_READABLE);
 
+    /**
+     * WebKitWebView:has-immersive-session:
```
This could have an introspection attribute to link the getter function to the property:

```suggestion
     * WebKitWebView:has-immersive-session: (attributes org.gtk.Property.get=webkit_web_view_get_has_immersive_session) (getter get_has_immersive_session):
```

(Yes, I know most of WebKit code lacks this kind of GI annotations, but that's because GTK-Doc&mdash;used in the past&mdash;didn't support them... it's always a good moment to start using them!)

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5866,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+}
+
+/**
+ * webkit_web_view_get_has_immersive_session:
```
Conversely, here we can indicate the property fetched by this getter:

 ```suggestion
 * webkit_web_view_get_has_immersive_session: (get-property has-immersive-session):
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5866,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ * own based on application or platform logic.
+ *
+ * Note that if WebXR is disabled, of if it is enabled but no immersive session
+ * exists,his API does nothing. See also webkit_web_view_get_has_immersive_session()
```
Small typo:

```suggestion
 * exists, this API does nothing. See also webkit_web_view_get_has_immersive_session()
```

aperezdc reviewed (approved):
LGTM, I left a couple of notes about the documentation comments.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5865,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ * own based on application or platform logic.
+ *
+ * Note that if WebXR is disabled, of if it is enabled but no immersive session
+ * exists, this API does nothing. See also webkit_web_view_get_has_immersive_session()
```
```suggestion
 * exists, this API does nothing. See also webkit_web_view_get_has_immersive_session().
```

mcatanzaro reviewed (commented):
Looks good, thanks.

It does need an API test, but no need to overthink this. The test can be simple. I understand WebXR itself is probably hard to test via API tests, but all we need to test here is that the APIs you've added work properly: e.g. check that the property and getter are both FALSE initially, start an XR session using JavaScript, check that they're both TRUE, end the session using your new API, check that they're both FALSE. Should be simple?

carlosgcampos commented on Source/WebKit/UIProcess/API/APIUIClient.h:
```diff
@@ -239,6 +239,9 @@ class UIClient {
     virtual void endXRSession(WebKit::WebPageProxy&, WebKit::PlatformXRSessionEndReason) { }
+#elif USE(OPENXR)
+    virtual void startXRSession(WebKit::WebPageProxy&) { }
+    virtual void endXRSession(WebKit::WebPageProxy&) { }
```
These are not the same as the iOS ones. I don't know how it works in iOS but I guess the session is handled by the browser, so this is used to create and end the session. In our case they are just notifications, so I think they should be called didStartXRSession and didEndXRSession.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -247,6 +247,7 @@ enum {
     PROP_DEFAULT_CONTENT_SECURITY_POLICY,
 
     PROP_THEME_COLOR,
+    PROP_HAS_IMMERSIVE_SESSION,
```
Just curiosity, why are we avoiding using XR in the API names? The spec is WebXR, for me it's easier to associate XR_SESSION with the spec than IMMERSIVE_SESSION

carlosgcampos replied:
I guess we only support sessions with an immersive mode?

spenap replied:
> I guess we only support sessions with an immersive mode?

Thanks for challenging this. At the moment, even if the property mentions _immersive_, it also reports when an inline XR session has started or has ended, and the same goes for `end_immersive_session`: it affects to both _immersive_ and _inline_ sessions.

So your earlier suggestion sounds right: I can rename our usage of _immersive_ with _xr_ so it covers both scenarios.

carlosgcampos replied:
Are you sure we support inline sessions? Because in that case it's not only a matter of renaming, since we are setting the property to false whenever one session ends, but there might be others active. Looking at the code, I think we only support immersive sessions, but I'm not sure.

spenap replied:
> Are you sure we support inline sessions? Because in that case it's not only a matter of renaming, since we are setting the property to false whenever one session ends, but there might be others active. Looking at the code, I think we only support immersive sessions, but I'm not sure.

sorry, no, my bad. I was thinking of the examples here, where there's an inline one: https://immersive-web.github.io/webxr-samples/

But it's only when we enter an _immersive_ session that this API is involved. So discard my earlier comment.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -1748,6 +1753,20 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
+     */
+    sObjProperties[PROP_HAS_IMMERSIVE_SESSION] =
+    g_param_spec_boolean(
+        "has-immersive-session",
```
I think we normally use `is-` fot this kind of properties, this could be `is-immersive-session-active` or something like that

aperezdc replied:
I actually looked a bit at this because it was also my impression, but we do have getter methods that use “has” as a prefix, like [webkit_credential_has_password()](https://wpewebkit.org/reference/stable/wpe-webkit-2.0/method.Credential.has_password.html) or  [wpe_view_has_focus()](https://people.igalia.com/aperez/Documentation/wpe-platform-2.0/method.View.get_has_focus.html) (and this one even has a `WPEView:has-focus` property along). I also checked other GObject APIs and there are more examples like `GdkDevice:has-cursor`, `GtkComboBox:has-entry`, `GtkWidget:has-default` and so on.

Therefore, I think it better to use the `has-` prefix here instead of trying to find some awkward wording that makes the property name longer than it needs to be.

carlosgcampos replied:
Yes, the problem is not using has in general, but how it's used in this particular case. If we only support immersive sessions and there can only be one active immersive session, maybe we can avoid using the word session. When the session is created we are effectively entering immersive mode and when the session is destroyed we are leaving it. This is similar to the pointer lock or fullscreen apis, the view enters/leave a mode, it's not something it has, but a mode it's enabled disabled. And in those cases we already leave pointer lock and fullscreen when Esc is pressed, so if we are exposing that api only for that use case, I think it's better to handle it internally.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5865,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ *
+ * Returns: %TRUE if there is an immersive session associated to the @web_view, or %FALSE otherwise.
+ *
+ * Note that if WebXR is disabled, this API always returns %FALSE.
```
Move this before the Returns. This comment would be more obvious if the property is xr-session instead of immersive session, no?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5865,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ * end the sessions themselves, but applications might need to end a session on their
+ * own based on application or platform logic.
+ *
+ * Note that if WebXR is disabled, of if it is enabled but the @web_view is not in
```
```suggestion
 * Note that if WebXR is disabled, or if it is enabled but the @web_view is not in
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
```diff
@@ -374,6 +374,16 @@ class UIClient : public API::UIClient {
+    {
+        webkitWebViewSetIsImmersiveModeEnabled(m_webView, false);
+    }
+
```
This is defined only when ENABLE(WEBXR) we should add guards here

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -428,6 +429,7 @@ struct _WebKitWebViewPrivate {
     WebKitWebExtensionMode webExtensionMode;
 
     bool isWebProcessResponsive;
+    bool isImmersiveModeEnabled;
```
This could be guarded too

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5201,6 +5220,7 @@ void webkitWebViewWebProcessTerminated(WebKitWebView* webView, WebKitWebProcessT
 
     // Reset the state of the responsiveness property.
     webkitWebViewSetIsWebProcessResponsive(webView, true);
+    webkitWebViewSetIsImmersiveModeEnabled(webView, false);
```
This too and I would add an empty line or update the comment, because it looks like it applies to both.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5865,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+    webView->priv->isImmersiveModeEnabled = isImmersiveModeEnabled;
+    g_object_notify_by_pspec(G_OBJECT(webView), sObjProperties[PROP_IS_IMMERSIVE_MODE_ENABLED]);
+}
+
```
This also needs #ifdef

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5865,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+{
+    g_return_val_if_fail(WEBKIT_IS_WEB_VIEW(webView), FALSE);
+
+    return webView->priv->isImmersiveModeEnabled;
```
And this, just return FALSE when WEBXR is not enabled in build.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5865,62 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+    if (!webView->priv->isImmersiveModeEnabled)
+        return;
+
 #if ENABLE(WEBXR)
```
And this should be moved before the if

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5870,67 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+
+    webView->priv->isImmersiveModeEnabled = isImmersiveModeEnabled;
+    g_object_notify_by_pspec(G_OBJECT(webView), sObjProperties[PROP_IS_IMMERSIVE_MODE_ENABLED]);
+#endif
```
Since we have guards for every cal tot this, it's better to define it only for WEBXR too and we avoid warnings for unused parameters when it's disabled.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
```diff
@@ -132,3 +132,5 @@ guint createContextMenuSignal(WebKitWebViewClass*);
 WebKit::RendererBufferDescription webkitWebViewGetRendererBufferDescription(WebKitWebView*);
 #endif
+
+void webkitWebViewSetIsImmersiveModeEnabled(WebKitWebView*, bool);
```
```suggestion
#if ENABLE(WEBXR)
void webkitWebViewSetIsImmersiveModeEnabled(WebKitWebView*, bool);
#endif
```

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,152 @@
+"});"
+"</script></body></html>";
+
+class ImmersiveModeTest : public ConsoleMessageTest {
```
I don think it's a good idea to do this, this is not a console message test. Instead of using console messages you can use user script messages. In other tests we also use the title for simple cases like this.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,152 @@
+#if !ENABLE(2022_GLIB_API)
+        WebKitWebsiteDataManager* m_manager;
+#endif
+    };
```
This is a great idea, if you move this to WebViewTest it could be used from TestSSL.cpp

spenap replied:
I've created a separate PR https://github.com/WebKit/WebKit/pull/49215

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,152 @@
+static void testWebKitImmersiveModeLeaveImmersiveMode(ImmersiveModeTest* test, gconstpointer)
+{
+    auto httpsServer = makeUnique<WebKitTestServer>(WebKitTestServer::ServerHTTPS);
+    httpsServer->run(serverCallback);
```
This is ok because we only have one test, but we normally make the server global, created in beforeAll and deleted in afterAll

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,152 @@
+    ConsoleMessageTest::ConsoleMessage referenceMessage = { ConsoleMessageTest::MessageSource::ConsoleAPI, ConsoleMessageTest::MessageLevel::Error, "XR session failed to start: SecurityError: The operation is insecure.", 1, httpsServer->getURIForPath("/xr-session/").data() };
+#else
+    ConsoleMessageTest::ConsoleMessage referenceMessage = { ConsoleMessageTest::MessageSource::JavaScript, ConsoleMessageTest::MessageLevel::Error, "TypeError: undefined is not an object (evaluating 'navigator.xr.requestSession')", 1, httpsServer->getURIForPath("/xr-session/").data() };
+#endif
```
I don't think we want to test this, just make the whole test guarded by ENABLE(WEBXR)

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,152 @@
+    test->runJavaScriptAndWaitUntilFinished("document.getElementById('enterXR').click();", &error.outPtr(), nullptr);
+    g_assert_no_error(error.get());
+#if ENABLE(WEBXR)
+    ConsoleMessageTest::ConsoleMessage referenceMessage = { ConsoleMessageTest::MessageSource::ConsoleAPI, ConsoleMessageTest::MessageLevel::Error, "XR session failed to start: SecurityError: The operation is insecure.", 1, httpsServer->getURIForPath("/xr-session/").data() };
```
I don't understand the test, this is an api test, we want to check that our property changes when a session is created and that we can leave the session from the api. If we can't test that, it doesn't make sense to add an api test. To check that requestSession gives a security error we can use a layout test instead.

spenap replied:
We can't enter a session programmatically, because click doesn't focus the document.

If there is a way to focus the document, then we can enter the session, and then we can also exit it successfully

spenap replied:
I commented out the focus requirement (because it's no longer present in the upstream version of the spec), and then the test can be more interesting.

I am not sure we can relax the webxr condition though, from reading the spec it looks like not asking for document focus is covered by an earlier check. 

I'll share my findings in the topic on Monday

spenap replied:
However, this test is an answer to @mcatanzaro ask for a simple test. If this doesn't cut it, I'm ok removing the test but still submitting the API changes, at least until we have a mock XR coordinator where a test could return.

carlosgcampos replied:
We are doing programmatic clicks in other tests, you need to ensure `test->showInWindow()` is called and then you can use `test->clickMouseButton()`

spenap replied:
I got this working by combining

```
        showInWindow();
        runJavaScriptAndWaitUntilFinished("document.getElementById('enterXR').focus()", nullptr);
        runJavaScriptAndWaitUntilFinished("document.getElementById('enterXR').click();", nullptr);
```

With that, I can enter an XR session **if Monado is running**. But if it isn't running, then the session simply doesn't start, so we can't test much -- only that the exception is thrown in JS.

fujii commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5845,25 +5870,67 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+    webView->priv->isImmersiveModeEnabled = isImmersiveModeEnabled;
+    g_object_notify_by_pspec(G_OBJECT(webView), sObjProperties[PROP_IS_IMMERSIVE_MODE_ENABLED]);
+#endif
+}
```
`#endif` should be placed after `}` here.
~~~
}
#endif
~~~

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+static void testWebKitImmersiveModeLeaveImmersiveMode(ImmersiveModeTest* test, gconstpointer)
+{
+    if (!g_getenv("WITH_OPENXR_RUNTIME")) {
+        g_test_skip("Not running without an OpenXR runtime");
```
I guess this is "Not running with" or "Running without"

spenap replied:
No, this is fine. If the environment variable is not found, then we skip the test because we *can't* run without an OpenXR runtime.

carlosgcampos replied:
I understand that, but the reason why we skip the test is because we are "Running without an OpenXR runtime " or "Not running with an OpenXR runtime", no?

spenap replied:
Ah! I think the problem is what we understood as the implied subject: you are thinking of the test runner and I was thinking of the individual test.

I read this as "[Test] not running without an OpenXR runtime", and you understand it as "[Test runner] running without an OpenXR runtime". 

I am happy to change it. Do you have any preference?

carlosgcampos replied:
I still don't understand it, to be honest. Even if it's the test, the reason for skipping is either test is running without an OpenXR runtime or test is not running with an OpenXR runtime. The double negation doesn't sound right to me, but I'm not native english.

spenap replied:
The test is "refusing to run without (e.g. not running) without an OpenXR runtime". It's like "I'm not talking without my lawyer". It doesn't matter, though.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+        soup_message_body_append(responseBody, SOUP_MEMORY_STATIC, indexHTML, strlen(indexHTML));
+        soup_message_body_complete(responseBody);
+    } else {
+        g_assert(false);
```
What is this?

spenap replied:
Do you mean "if we are not planning to use paths other than `"/xr-session/"` we don't need to handle them in the callback" or is it a different question?

carlosgcampos replied:
g_assert_not_reached()

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+
+    void simulateUserInteraction()
+    {
+        showInWindow();
```
Better to this at the beginning of the test

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+        webkit_web_view_leave_immersive_mode(m_webView.get());
+    }
+
+    void simulateUserInteraction()
```
I would call this clickEnterXRButton. We use simulateUserInteraction in UI client tests to simulate user interaction for things that require it, but here we are actually clicking the button we want.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+    kHttpsServer = new WebKitTestServer(WebKitTestServer::ServerHTTPS);
+    kHttpsServer->run(serverCallback);
+
+    Test::shouldInitializeWebProcessExtensions = true;
```
Why do we need this?

spenap replied:
I think I had that in an earlier attempt to get JS notifications -- before I had the console, and before I implemented the property and could replace the console with the notify:: 

I'll remove it

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+    test->simulateUserInteraction();
+
+    if (!test->isImmersiveModeEnabled())
+        test->waitUntilIsImmersiveModeEnabledChanged();
```
These could be just one function test->clickOnEnterXRButtonAndWaitUntilImmersiveModeChanged().

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+    void leaveImmersiveMode()
+    {
+        webkit_web_view_leave_immersive_mode(m_webView.get());
+    }
```
I prefer not to use wrappers for single line calls to the api, because when reading the test it's not so obvious what api we are testing.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitImmersiveMode.cpp:
```diff
@@ -0,0 +1,138 @@
+    test->leaveImmersiveMode();
+
+    if (test->isImmersiveModeEnabled())
+        test->waitUntilIsImmersiveModeEnabledChanged();
```
leaveImmersiveModeAndWaitUntilChanged()

spenap committed:
[GTK][WPE][OpenXR] Expose a way to query the immersive session status
https://bugs.webkit.org/show_bug.cgi?id=297048

Reviewed by Carlos Garcia Campos and Adrian Perez de Castro.

A new property "is-immersive-mode-enabled" is added to WebKitWebView,
allowing clients to query its value and to subscribe to its notify
signal.

The property is synchronised with the OpenXR coordinator's entering and
exiting the session, where the WebPageProxy's UI client is used to
report back to the WebView.

This renames webkit_web_view_end_immersive_session into
webkit_web_view_leave_immersive_mode for better consistency.

* Source/WebKit/UIProcess/API/APIUIClient.h: two new methods to report
  entering and exiting the XR session.
(API::UIClient::didStartXRSession):
(API::UIClient::didEndXRSession):
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp: implementing the
  UI client methods in terms of updating the immersive mode enabled
property.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: add the new
  property, rename method leaving the mode.
(webkitWebViewGetProperty):
(webkit_web_view_class_init):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in: rename method
  leaving the mode.
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h: adding method
  setting the immersive mode enabled property.
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp: report when we
  enter and leave the XR session.
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
* Tools/MiniBrowser/wpe/main.cpp: query whether we're in immersive mode
  before attempting to leave.

Canonical link: https://commits.webkit.org/298568@main
